lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -1744,23 +1751,90 @@ pub struct CGOSharedDirectoryMoveRequest {
     pub new_path: *const c_char,
 }
 
-pub type CGOSharedDirectoryCreateResponse = SharedDirectoryCreateResponse;
+/// SharedDirectoryCreateResponse is sent by the TDP client to the server
+/// to acknowledge a SharedDirectoryCreateRequest was received and executed.
+#[derive(Debug)]
+pub struct SharedDirectoryCreateResponse {
+    completion_id: u32,
+    err_code: TdpErrCode,
+    fso: FileSystemObject,
+}
+
+#[repr(C)]
+pub struct CGOSharedDirectoryCreateResponse {
+    pub completion_id: u32,
+    pub err_code: TdpErrCode,
+    pub fso: CGOFileSystemObject,
+}
+
+impl From<CGOSharedDirectoryCreateResponse> for SharedDirectoryCreateResponse {
+    fn from(cgo_res: CGOSharedDirectoryCreateResponse) -> SharedDirectoryCreateResponse {
+        // # Safety
+        //
+        // This function MUST NOT hang on to any of the pointers passed in to it after it returns.
+        // In other words, all pointer data that needs to persist after this function returns MUST
+        // be copied into Rust-owned memory.
+        SharedDirectoryCreateResponse {
+            completion_id: cgo_res.completion_id,
+            err_code: cgo_res.err_code,
+            fso: FileSystemObject::from(cgo_res.fso),
+        }
+    }
+}
+
 /// SharedDirectoryDeleteRequest is sent by the TDP server to the client
 /// to request the deletion of a file or directory at path.
-pub type SharedDirectoryDeleteRequest = SharedDirectoryInfoRequest;
-pub type CGOSharedDirectoryDeleteRequest = CGOSharedDirectoryInfoRequest;
+#[derive(Debug)]
+pub struct SharedDirectoryDeleteRequest {
+    completion_id: u32,
+    directory_id: u32,
+    path: UnixPath,
+}
+
+#[repr(C)]
+pub struct CGOSharedDirectoryDeleteRequest {
+    pub completion_id: u32,
+    pub directory_id: u32,
+    pub path: *const c_char,
+}
+
 /// SharedDirectoryDeleteResponse is sent by the TDP client to the server
 /// to acknowledge a SharedDirectoryDeleteRequest was received and executed.
-pub type SharedDirectoryDeleteResponse = SharedDirectoryCreateResponse;
-pub type CGOSharedDirectoryDeleteResponse = SharedDirectoryCreateResponse;
-/// SharedDirectoryListRequest is sent by the TDP server to the client
-/// to request the contents of a directory.
-pub type SharedDirectoryListRequest = SharedDirectoryInfoRequest;
-pub type CGOSharedDirectoryListRequest = CGOSharedDirectoryInfoRequest;
+#[derive(Debug)]
+#[repr(C)]
+pub struct SharedDirectoryDeleteResponse {
+    completion_id: u32,
+    err_code: TdpErrCode,
+}
+
+pub type CGOSharedDirectoryDeleteResponse = SharedDirectoryDeleteResponse;
+
 /// SharedDirectoryMoveResponse is sent by the TDP client to the server
 /// to acknowledge a SharedDirectoryMoveRequest was received and expected.
-pub type SharedDirectoryMoveResponse = SharedDirectoryCreateResponse;
-pub type CGOSharedDirectoryMoveResponse = CGOSharedDirectoryCreateResponse;
```

Comment:
I realized that all of these inter-message dependencies was a bad idea, since it means changing an individual message metaphorically pulls a thread that can lead to all sorts of other messages being mistakenly changed. Now each message has its own separate `struct`.